notifier/service/notifier.go

```diff
@@ -0,0 +1,216 @@
+	AMQP             *config.AMQP
+	STOMP            *config.STOMP
+	PollInterval     time.Duration
+	DeliveryInterval time.Duration
+	DisableSummary   bool
+}
+
+// New returns a configured notifier subsystem.
+func New(ctx context.Context, store notifier.Store, locks notifier.Locker, opts Opts) (*Notifier, error) {
+	ctx = baggage.ContextWithValues(ctx,
+		label.String("component", "notifier/service/New"),
+	)
+	srv := Notifier{store: store}
+
+	// Check for test mode.
+	if tm := os.Getenv("NOTIFIER_TEST_MODE"); tm != "" {
+		zlog.Warn(ctx).
+			Stringer("interval", opts.PollInterval).
+			Msg("NOTIFIER TEST MODE ENABLED. NOTIFIER WILL CREATE TEST NOTIFICATIONS ON A SET INTERVAL")
+		testModeInit(ctx, &opts)
+	}
+
+	// Configure the Poller.
+	zlog.Info(ctx).
+		Stringer("interval", opts.PollInterval).
+		Msg("initializing poller")
+	srv.poll = notifier.NewPoller(store, opts.Matcher, opts.PollInterval)
+
+	// Configure the Processor.
+	zlog.Info(ctx).
+		Int("count", processors).
+		Msg("initializing processors")
+	srv.proc = notifier.NewProcessor(store, locks, opts.Indexer, opts.Matcher)
+	srv.proc.NoSummary = opts.DisableSummary
+
+	// Configure a Deliverer.
+	var del notifier.Deliverer
+	var err error
+	// BUG(hank) Currently only one delivery mechanism can be configured at a
+	// time.
+	switch {
+	case opts.Webhook != nil:
+		zlog.Info(ctx).
+			Int("count", deliveries).
+			Msg("initializing webhook deliverers")
+		del, err = webhook.New(opts.Webhook, opts.Client)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create webhook deliverer: %v", err)
+		}
+	case opts.AMQP != nil:
+		conf := opts.AMQP
+		if len(conf.URIs) == 0 {
+			zlog.Warn(ctx).
+				Msg("amqp delivery misconfigured: no broker URIs to connect to")
+			break
+		}
+		if conf.Direct {
+			del, err = amqp.NewDirectDeliverer(conf)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create AMQP deliverer: %v", err)
+			}
+			break
+		}
+		del, err = amqp.New(conf)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create AMQP deliverer: %v", err)
+		}
+	case opts.STOMP != nil:
+		conf := opts.STOMP
+		if len(conf.URIs) == 0 {
+			zlog.Warn(ctx).
+				Msg("stomp delivery misconfigured: no broker URIs to connect to")
+			break
+		}
+		if conf.Direct {
+			del, err = stomp.NewDirectDeliverer(conf)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create STOMP direct deliverer: %v", err)
+			}
+			break
+		}
+		del, err = stomp.New(conf)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create STOMP deliverer: %v", err)
+		}
+	}
+	if del == nil {
+		// Report an error if configured such that no notifications are being
+		// processed.
+		return nil, ErrNoDelivery
+	}
+	srv.del = notifier.NewDelivery(store, locks, del, opts.DeliveryInterval)
+
+	return &srv, nil
+}
+
+// TestModeInit will inject a mock Indexer and Matcher into opts
+// to be used in testing mode.
+func testModeInit(ctx context.Context, opts *Opts) error {
```

Comment:
This might be nicer if it returned `(*Opts, error)` what do you think?

Reply:
:man_shrugging: I think this way of triggering the test behavior is a giant wart that needs to get fixed anyway.